cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -485,6 +485,106 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
+                            // A' = Ah' * Bl
+                            // B' = B
+                            // B' = B' >> 32
+                            // B' = Bh' * Al
+                            // B' = B' + A'
+                            // B' = B' << 32
+                            // dst = Al * Bl
+                            // dst = dst + B'
+
+                            // Get inputs rhs = A and lhs = B and the dst register
+                            let lhs = put_input_in_reg(ctx, inputs[0]);
+                            let rhs = put_input_in_reg(ctx, inputs[1]);
+                            let dst = get_output_reg(ctx, outputs[0]);
+
+                            // A' = A
+                            let rhs_1 = ctx.alloc_tmp(RegClass::V128, types::I64X2);
+                            ctx.emit(Inst::gen_move(rhs_1, rhs, ty));
+
+                            // A' = A' >> 32
+                            // A' = Ah' * Bl
+                            ctx.emit(Inst::xmm_rmi_reg(
+                                SseOpcode::Psrlq,
+                                RegMemImm::imm(32),
+                                rhs_1,
+                            ));
+                            ctx.emit(Inst::xmm_rm_r(
+                                SseOpcode::Pmuludq,
+                                RegMem::reg(lhs.clone()),
+                                rhs_1,
+                            ));
+
+                            // B' = B
+                            let lhs_1 = ctx.alloc_tmp(RegClass::V128, types::I64X2);
+                            ctx.emit(Inst::gen_move(lhs_1, lhs, ty));
+
+                            // B' = B' >> 32
+                            // B' = Bh' * Al
+                            ctx.emit(Inst::xmm_rmi_reg(
+                                SseOpcode::Psrlq,
+                                RegMemImm::imm(32),
+                                Writable::from_reg(lhs_1.to_reg()),
+                            ));
+                            ctx.emit(Inst::xmm_rm_r(
+                                SseOpcode::Pmuludq,
+                                RegMem::reg(rhs),
+                                Writable::from_reg(lhs_1.to_reg()),
+                            ));
+
+                            // B' = B' + A'
+                            // B' = B' << 32
+                            ctx.emit(Inst::xmm_rm_r(
+                                SseOpcode::Paddq,
+                                RegMem::reg(rhs_1.to_reg()),
+                                lhs_1,
+                            ));
+                            ctx.emit(Inst::xmm_rmi_reg(
+                                SseOpcode::Psllq,
+                                RegMemImm::imm(32),
+                                lhs_1,
+                            ));
+
+                            // dst = Al * Bl
+                            // dst = dst + B'
+                            ctx.emit(Inst::gen_move(dst, rhs, ty));
```

Comment:
I think I was the one who suggested using the dst register as a temporary, but thinking about it: sorry, we can't do this. It might be that `lhs` will get the same physical register as `dst` (if `lhs`'s final use is for this instruction), thus this move would clobber `lhs` before it's used in the next `pmuludq` instruction. Instead, you'll need to use a temporary in place of `dst` here and in the next instruction; I think you can reuse `rhs_1` though, since it's dead at this point. (This might require a final `gen_move(dst, rhs_1, ty))` after the whole sequence, which is fine too.)

Reply:
Makes sense. Thanks!